cmd/gardenlet/app/app.go

```diff
@@ -451,6 +465,133 @@ func (g *garden) Start(ctx context.Context) error {
+			grmDeploy := appsv1.Deployment{}
+			if err := seedClient.Get(ctx, types.NamespacedName{Namespace: ns.Name, Name: "gardener-resource-manager"}, &grmDeploy); err != nil {
+				if apierrors.IsNotFound(err) {
+					// Do not proceed if GRM deployment is not present
+					return nil
+				}
+				return err
+			}
+
+			// Create a DWDAccess object
+			InClusterServerURL := fmt.Sprintf("%s.%s.svc", v1beta1constants.DeploymentNameKubeAPIServer, ns.Name)
+			dwdAccess := dwd.NewDWDAccess(seedClient, ns.Name, nil, dwd.AccessValues{ServerInCluster: InClusterServerURL})
+
+			err := dwdAccess.DeployMigrate(ctx)
+			if err != nil {
+				return err
+			}
+
+			// Fetch and update the GRM configmap
+			grmCM := corev1.ConfigMap{}
+			volumeNameConfig := "config"
+			grmCMName := ""
+			for _, vol := range grmDeploy.Spec.Template.Spec.Volumes {
+				if vol.Name == volumeNameConfig {
+					grmCMName = vol.ConfigMap.Name
+				}
+			}
+			if len(grmCMName) == 0 {
+				return nil
+			}
+			if err := seedClient.Get(ctx, types.NamespacedName{Namespace: ns.Name, Name: grmCMName}, &grmCM); err != nil {
+				if apierrors.IsNotFound(err) {
+					return nil
+				}
+				return err
+			}
+
+			cmData := grmCM.Data["config.yaml"]
+			rmConfig := resourcemanagerv1alpha1.ResourceManagerConfiguration{}
+
+			// create codec
+			var codec runtime.Codec
+			configScheme := runtime.NewScheme()
+			utilruntime.Must(resourcemanagerv1alpha1.AddToScheme(configScheme))
+			utilruntime.Must(apiextensionsv1.AddToScheme(configScheme))
+			ser := json.NewSerializerWithOptions(json.DefaultMetaFactory, configScheme, configScheme, json.SerializerOptions{
+				Yaml:   true,
+				Pretty: false,
+				Strict: false,
+			})
+			versions := schema.GroupVersions([]schema.GroupVersion{
+				resourcemanagerv1alpha1.SchemeGroupVersion,
+				apiextensionsv1.SchemeGroupVersion,
+			})
+			codec = serializer.NewCodecFactory(configScheme).CodecForVersions(ser, ser, versions, versions)
+
+			obj, err := runtime.Decode(codec, []byte(cmData))
+			if err != nil {
+				return err
+			}
+			rmConfig = *(obj.(*resourcemanagerv1alpha1.ResourceManagerConfiguration))
+
+			rmConfig.TargetClientConnection.Namespaces = append(rmConfig.TargetClientConnection.Namespaces, corev1.NamespaceNodeLease)
+
+			data, err := runtime.Encode(codec, &rmConfig)
+			if err != nil {
+				return err
+			}
+
+			err = seedClient.Delete(ctx, &grmCM)
+			if err != nil {
+				return err
+			}
+
+			grmCM.Data = map[string]string{"config.yaml": string(data)}
+			grmCM.ObjectMeta.Name = dwd.TempGRMConfigMapName
+			grmCM.ObjectMeta.ResourceVersion = ""
+
+			err = seedClient.Create(ctx, &grmCM)
+			if err != nil {
+				return err
+			}
+
+			// Update GRM delpoy to reference the newly created cm
+			grmDeployCopy := grmDeploy.DeepCopy()
+			for n, vol := range grmDeploy.Spec.Template.Spec.Volumes {
+				if vol.Name == "config" {
+					grmDeploy.Spec.Template.Spec.Volumes[n].ConfigMap.Name = dwd.TempGRMConfigMapName
+				}
+			}
```

Comment:
Why don't you save the index of the volume in the prior loop and use it to patch the name? Then you don't have to iterate here again.